[Share 2.0] register providers

rullzer opened the pull request:
For #19331

Allow registering of providers.

Lets see what CI has to say.

rullzer committed:
[Share 2.0] Allow registering of share providers

* Properly register the default share provider

DeepDiver1975 commented:
Will there ever be the need to have multiple providers of the same type?

If not we can move this to pure config as we did this with the commants manager
https://github.com/owncloud/core/blob/master/lib/private/comments/managerfactory.php
https://github.com/owncloud/core/blob/master/config/config.sample.php#L812

Would result in less code - and less to test ....

rullzer commented:
@DeepDiver1975 no I don't think so. Because if there are multiple providers for the same type how will we know which once to use.

We will need multiple providers. And multiple types per provider.

Moving to config is also fine with me...

@schiesbn opinion?

DeepDiver1975 commented:
I don't think we need mutliple providers for the same share type.

From my POV this is beyond the current concept to have 10 implementations of public link sharing running at the same time.

rullzer commented:
@DeepDiver1975 yeah I'm agreeing with you :P

schiesbn commented:
I don't think we need to have multiple providers for the same share type. I would only allow to register one provider for one share type. Even if, at some point in time, we have two share provider which do some kind of link sharing they should have a different share type. For example "linkShare" and "SuperLinkShareNG". The share type should the the unique identifier which connects a share to a share provider.

DeepDiver1975 commented:
so - reading the classname from the config is much easier from my pov

rullzer commented:
Ok let me look into that then :)

rullzer commented:
Ok so after looking into it. I think it having this into config.php makes sense. Since it requires explicit action from the admin to setup the different share providers. Which I think is a good thing here.

But since we will have 2 share providers at least (default and federated) I'm thinking how to represent this. And how not to complecate the setup to much. Because we need to know for which share types.

So in short the code needs to stay I think. But it reduces stuff in the public name space.

DeepDiver1975 commented:
> But since we will have 2 share providers at least (default and federated) I'm thinking how to represent this. And how not to complecate the setup to much. Because we need to know for which share types.

I'd leave this to the factory - the factory is the class which is defined in the config.

The factory method will return the provider for the asked share type.

rullzer commented:
@DeepDiver1975 you mean that we list in general 2 provider factories there.
Then we call those factories and they return the class and an array of share types which wee will then register at the manager?

I can see that working :)

DeepDiver1975 commented:
One factory to rule them all.

rullzer commented:
> One factory to rule them all. 

But if people want to deploy their own sharae provider it is likely they only want to do the 'main' share provider. But keep the federated sharing for example. Would they then need to create their own factory to create their own provider and the federated provider?

rullzer committed:
[Share 2.0] Add share provider factory

* Add providers
* Add share manager to server container
* Use share manager from server container
* Properly get the share manager

rullzer commented:
Now using the factory apporoach. Everything still works as expected.

nickvergessen commented on lib/private/share20/iproviderfactory.php:
```diff
@@ -0,0 +1,32 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+namespace OC\Share20;
```
should  this be public?

rullzer replied:
yes, eventually. But I want to move that over once I also move the other interfaces to OCP

DeepDiver1975 commented on lib/private/server.php:
```diff
@@ -564,6 +564,26 @@ public function __construct($webRoot, \OC\Config $config) {
+				$c->getL10N('core')
+			);
+
+			$manager->addProviders($factory->getProviders());
```
hmmm . this would result in creating all providers up front - using the factory within the manager and ask for the explicit provider as needed would have been my approach

rullzer replied:
Sure we can do that as well of course. However we need to create all providers at the same time. But that would still be better of course since then it only happens when we actually want to do something with sharing.

rullzer replied:
Let me fix this :)

rullzer commented:
Ok now we call the factory only in the manager. There might be more room for optimization there. But I'd like to move forward with this PR.

CC: @DeepDiver1975 @nickvergessen

LukasReschke commented on lib/private/share20/iproviderfactory.php:
```diff
@@ -0,0 +1,32 @@
+	 * @return IShareProvider[]
+	 */
+	public function getProviders();
+}
```
Newline at EoF

LukasReschke commented on lib/private/share20/manager.php:
```diff
@@ -79,13 +88,16 @@ class Manager {
 			IGroupManager $groupManager,
-			IL10N $l
+			IL10N $l,
+			IProviderFactory $factory
```
Missing in the PHPDoc

LukasReschke commented on lib/private/share20/manager.php:
```diff
@@ -93,9 +105,79 @@ public function __construct(
+	 * @param IShareProvider[] $providers
+	 * @throws ProviderException
+	 */
+	private function addProviders($providers) {
```
`array $providers`?

LukasReschke commented on lib/private/share20/providerfactory.php:
```diff
@@ -0,0 +1,50 @@
+
+		return $providers;
+	}
+}
```
Newline at Eof

LukasReschke commented on config/config.sample.php:
```diff
@@ -963,6 +963,22 @@
+ * own or 3rdParty Share Providers be used that – for instance – uses the
+ * filesystem instead of the database to keep the share information.
+ */
+'sharing.managerFactory' => '\OC\Share20\ProviderFactory',
```
I already see this exploding by all the ~~idiots~~ people that believe that copying our config.sample.php is the way to go, even if it says otherwise :cry:

But fair enough :smile:

rullzer replied:
Thinking about it some more. We might remove this from config.sample.php... 99.9999% of the people don't want to change it anyway...

Or maybe just comment it out by default.

rullzer replied:
Nah just leave it. Even if people copy the sample things should not break apart.

DeepDiver1975 replied:
we shall keep it since documentation is generated based on this file

LukasReschke commented on lib/private/server.php:
```diff
@@ -564,6 +564,25 @@ public function __construct($webRoot, \OC\Config $config) {
+		$this->registerService('ShareManager', function(Server $c) {
+			$config = $c->getConfig();
+			$factoryClass = $config->getSystemValue('sharing.managerFactory', '\OC\Share20\ProviderFactory');
+			/** @var \OC\Share20\IProviderFactory $factory */
```
:rocket:

LukasReschke commented on lib/private/share20/manager.php:
```diff
@@ -584,11 +678,32 @@ public function getShares($page=0, $perPage=50) {
+			throw new ShareNotFound();
+		}
 
 		return $share;
```
Makes static code analyzers sad:

![2016-01-13_07-59-34](https://cloud.githubusercontent.com/assets/878997/12287281/96f8faa4-b9cb-11e5-8367-1c8d3055d545.png)

While this seems not to be a real world scenario (no provider would need to be defined) it would maybe make sense to check `$found === false || empty($share)`? We can probably also get rid of `$found` completely and just check on `empty($share)`.

MorrisJobke replied:
I just also thought like this. Drop this `$found` ;)

nickvergessen replied:
:+1: for removing found and just defining `$share = false;` before the loop

LukasReschke commented:
:+1: besides my nitpicks :smile:

rullzer commented:
Comments from @LukasReschke addressed and rebased.

nickvergessen commented on lib/private/server.php:
```diff
@@ -1181,5 +1200,12 @@ public function getUserGlobalStoragesService() {
+
+
+	/**
+	 * @return \OC\Share20\Manager
```
should be public in the end, so apps can typehint the return

nickvergessen commented on lib/private/server.php:
```diff
@@ -1181,5 +1200,12 @@ public function getUserGlobalStoragesService() {
+	/**
+	 * @return \OC\Share20\Manager
+	 */
+	public function getShareManager() {
```
missing from the public interface

rullzer replied:
Yes for now. Until we move everything including the interfaces over..

nickvergessen commented on lib/private/share20/manager.php:
```diff
@@ -93,9 +106,79 @@ public function __construct(
+		foreach ($providers as $provider) {
+			$name = get_class($provider);
+
+			if (!($provider instanceof IShareProvider)) {
```
Would have been cool, if share providers could be closures, so they are not constructed when they are not needed, but then other stuff gets more complicated, so either way is fine by me.

rullzer replied:
Yes I agree. We can look at that in a later stage. For now I just want to get this in. But I agree this could be somewhat more elegant.

nickvergessen commented on lib/private/share20/manager.php:
```diff
@@ -93,9 +106,79 @@ public function __construct(
+	}
+
+	private function runFactory() {
+		if ($this->factoryDone === true) {
```
could use `!empty($this->providers)` here, and then get rid of the boolean flag?

rullzer replied:
Yes lets do that. Will fix.

nickvergessen commented on lib/private/share20/manager.php:
```diff
@@ -491,13 +577,21 @@ public function updateShare(Share $share) {
 
-			$this->defaultProvider->delete($child);
-			$deletedShares[] = $child;
+		$providerIds = array_keys($this->providers);
```
why?
```php
foreach($this->providers as $providerId => $provider) {
```

is what you actually want to use below, no need to circle over the keys and then call a method to get the value

nickvergessen replied:
Actually you don't even need the id, so
```php
foreach ($this->providers as $provider) {
```
and your done

nickvergessen commented on lib/private/share20/manager.php:
```diff
@@ -584,11 +679,33 @@ public function getShares($page=0, $perPage=50) {
+
+		$share = null;
+		foreach ($providerIds as $providerId) {
+			$provider = $this->getProvider($providerId);
```
same here again...
```php
foreach ($this->providers as $provider) {
```

nickvergessen commented on tests/lib/share20/managertest.php:
```diff
@@ -83,22 +89,82 @@ public function setUp() {
+	 * Initialise the factory. If $providers is null add the defaultProvider
+	 * @param array $providers
+	 */
+	private function facotryProviders(array $providers = null) {
```
faco what? :skull:

nickvergessen commented on lib/private/share20/manager.php:
```diff
@@ -588,7 +686,25 @@ public function getShareById($id) {
+			}
+		}
+
+		if ($found === false) {
```
`$share === null`
and then drop the found

DeepDiver1975 commented on lib/private/share20/manager.php:
```diff
@@ -93,9 +103,88 @@ public function __construct(
+	 * @throws ProviderException
+	 */
+	private function getProvider($id) {
+		$this->runFactory();
```
I'd simply forward this call to the factory - no need to build the array.

We need a provider -> ask the factory to build one and we use it.

DeepDiver1975 commented on lib/private/share20/manager.php:
```diff
@@ -492,12 +584,17 @@ public function updateShare(Share $share) {
-			$deletedShares[] = $child;
+		$providers = $this->getProviders();
+
+		foreach($providers as $provider) {
```
why do we need to iterate? the share has one share type - we can ask for the provider of the share type and delete the children in there. no need to loop

rullzer replied:
Ah yes, left over from the initial approach... but I agree that a share can only have children within its own provider... (actually it will only be used in our provider for user modified group shares).

DeepDiver1975 commented on lib/private/share20/manager.php:
```diff
@@ -588,7 +686,24 @@ public function getShareById($id) {
+		$providers = $this->getProviders();
+
+		$share = null;
+		foreach ($providers as $provider) {
```
same as above - no looping needed

rullzer replied:
here we need to loop. At least until we add the providerid. Which we can't do in the current situation since that would break OCS... but if I request a share with id 42. That could be 42 for the federated shares or 42 for the 'regular' shares

rullzer replied:
mmm I have an idea how to fix that... so this code is cleaner... let me look into that... but after that I want it merged :stuck_out_tongue_winking_eye:

rullzer committed:
[Share 2.0] Use full share id (providerId:shareId)

Now that we support multiple managers we communicate shares to the
outside as 'providerId:shareId'. This makes sures that id's are unique
when references from the OCS API.

However, since we do not want to break the OCS API v1 we need to
somewhat hack around this.

When we switch to OCS API v2 (which we should when we support more
custom providers). We will change the id to always be the fullShareId.

rullzer committed:
[Share 2.0] Let the factory do the factory stuff

* Updated unit tests (bit cleaner now)

rullzer commented:
Warg... this PR got bigger than intented again :(

Now:
* providers are all handled in the factory including fetching logic
* A share now has a fullId (which is 'providerId:shareId'). That way shares are unique. This is required for multiple share providers.
  - However the OCS endpoint v1 can't use them yet since there id's are ints. So we have to hack around it.

One more look please: @DeepDiver1975 @nickvergessen @schiesbn @LukasReschke

rullzer commented on lib/private/share20/providerfactory.php:
```diff
@@ -0,0 +1,78 @@
+	public function getProviderForType($shareType) {
+		//FIXME we should not report type 2
+		if ($shareType === \OCP\Share::SHARE_TYPE_USER  ||
+			$shareType === 2 ||
```
I will fix this in a future PR... doing it here will make it to big.

LukasReschke replied:
Mind to create an issue for this? - We all know what happens to FIXMEs in the code :wink:

DeepDiver1975 commented:
move it in :+1: 

any chance in squashing some of these commits?

rullzer commented:
> any chance in squashing some of these commits?

Done!

LukasReschke commented:
:pray: :+1:

DeepDiver1975 commented:
nice
